File: tracing.go
/*
	Copyright 2022 Loophole Labs

	Licensed under the Apache License, Version 2.0 (the "License");
	you may not use this file except in compliance with the License.
	You may obtain a copy of the License at

		   http://www.apache.org/licenses/LICENSE-2.0

	Unless required by applicable law or agreed to in writing, software
	distributed under the License is distributed on an "AS IS" BASIS,
	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
	See the License for the specific language governing permissions and
	limitations under the License.
*/

package scale

import (
	"context"

	"github.com/tetratelabs/wazero/api"
)

// getFunctionNameLen is the Host function for getting the Function Name Length
func (r *Scale[T]) getFunctionNameLen(_ context.Context, module api.Module, params []uint64) {
	r.activeModulesMu.RLock()
	m := r.activeModules[module.Name()]
	r.activeModulesMu.RUnlock()
	if m == nil {
		return
	}

	params[0] = uint64(len([]byte(m.function.identifier)))
}

// getFunctionName is the Host function for getting the Function Name
func (r *Scale[T]) getFunctionName(_ context.Context, module api.Module, params []uint64) {
	r.activeModulesMu.RLock()
	m := r.activeModules[module.Name()]
	r.activeModulesMu.RUnlock()
	if m == nil {
		return
	}

	ptr := uint32(params[0])
	mem := module.Memory()
	mem.Write(ptr, []byte(m.function.identifier))
}

// getInstanceID is the Host function to get 16 byte Instance ID
func (r *Scale[T]) getInstanceID(_ context.Context, module api.Module, params []uint64) {
	r.activeModulesMu.RLock()
	m := r.activeModules[module.Name()]
	r.activeModulesMu.RUnlock()
	if m == nil {
		return
	}

	ptr := uint32(params[0])
	mem := module.Memory()
	mem.Write(ptr, m.instance.instanceID)
}

// otelTraceJSON is the Host function to receive OTEL Trace data in JSON
// and then call the TraceDataCallback
func (r *Scale[T]) otelTraceJSON(_ context.Context, module api.Module, params []uint64) {
	if r.TraceDataCallback == nil {
		return
	}

	ptr := uint32(params[0])
	length := uint32(params[1])
	mem := module.Memory()
	if data, ok := mem.Read(ptr, length); ok {
		r.TraceDataCallback(string(data))
	}
}

File: tracing.ts
/*
        Copyright 2022 Loophole Labs

        Licensed under the Apache License, Version 2.0 (the "License");
        you may not use this file except in compliance with the License.
        You may obtain a copy of the License at

                   http://www.apache.org/licenses/LICENSE-2.0

        Unless required by applicable law or agreed to in writing, software
        distributed under the License is distributed on an "AS IS" BASIS,
        WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
        See the License for the specific language governing permissions and
        limitations under the License.
*/

import {Module} from "./module";
import {Signature} from "@loopholelabs/scale-signature-interfaces";

const ErrNoInstance = Error("no webassembly instance")
const ErrNoMemory = Error("no exported memory in webassembly instance")

export interface HostFunctions extends WebAssembly.ModuleImports {
  get_function_name_len(): number
  get_function_name(ptr: number): void
  get_instance_id(ptr: number): void
  otel_trace_json(ptr: number, len: number): void
}

export type CallbackFunction = (data: string) => void

export class Tracing<T extends Signature> {
  private exports: WebAssembly.Exports | undefined;

  private readonly module: Module<T>;
  private readonly traceCallback: CallbackFunction | undefined;

  constructor(module: Module<T>, traceCallback: CallbackFunction | undefined) {
    this.module = module;
    this.traceCallback = traceCallback;
  }

  getFunctionNameLen(): number {
    const enc = new TextEncoder();
    const data = enc.encode(this.module.template.identifier);
    return data.length;
  }

  getFunctionName(ptr: number) {
    const enc = new TextEncoder();
    const data = enc.encode(this.module.template.identifier);
    const buffer = this.getDataView();
    for (let i=0;i<data.length;i++) {
      const d = data.at(i);
      if (typeof d !== "undefined") {
        buffer.setInt8(ptr+i, d);
      }
    }
  }

  getInstanceId(ptr: number) {
    if (typeof this.module.function === "undefined") return;
    const buffer = this.getDataView();
    for (let i=0;i<this.module.function.instance.identifier.length;i++) {
      const d = this.module.function.instance.identifier.at(i);
      if (typeof d !== "undefined") {
        buffer.setInt8(ptr+i, d);
      }
    }
  }

  otelTraceJSON(ptr: number, len: number) {
    if (typeof this.traceCallback === "undefined") return;
    const buffer = this.getDataView();
    const data = buffer.buffer.slice(ptr, ptr + len);
    const dec = new TextDecoder();
    const s = dec.decode(data);
    this.traceCallback(s);
  }

  GetImports(): HostFunctions {
    return {
      get_instance_id: this.getInstanceId.bind(this),
      get_function_name_len: this.getFunctionNameLen.bind(this),
      get_function_name: this.getFunctionName.bind(this),
      otel_trace_json: this.otelTraceJSON.bind(this),
    }
  }

  private getDataView(): DataView {
    if (!this.exports) {
        throw ErrNoInstance;
    }
    if (!this.exports.memory) {
        throw ErrNoMemory;
    }

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    return new DataView(this.exports.memory.buffer);
  }
}